site/_posts/2020-08-07-locking-an-account-with-breached-password.md

```diff
@@ -0,0 +1,258 @@
+
+```php
+//...
+$input = file_get_contents('php://input');
+
+$obj = json_decode($input);
+//...
+```
+
+In these lines, the entire contents of the webhook payload are converted into a string. The string is then decoded into a JSON object for easier handling.
+
+```php
+//...
+if (!$obj) { 
+  error_log("Invalid JSON");
+  return;
+}
+
+type = $obj->event->type;
+if ($type !== "user.password.breach") {
+  error_log("Sorry, we only handle breached password events.");
+  return;
+}
+
+$user_id = $obj->event->user->id;
+if (!$user_id) {
+  error_log("No user id");
+  return;
+}
+//...
+```
+
+Next, validate the payload. If you don't get valid JSON, a password breach event, and a user id, simply log an error and return. Doing so allows the webhook and the login event to succeed. Check for the event type just in case there's a misconfiguration and our webhook is notified of other types of events.
+
+```php
+//...
+$client = new FusionAuth\FusionAuthClient($api_key, $fa_url);
+$response = $client->deactivateUser($user_id);
+if (!$response->wasSuccessful()) {
+  error_log("Response wasn't successful:");
+  error_log(var_export($response, TRUE));
+  return;
+}
+//...
+```
+
+Finally! This is where the action is. This code creates a new FusionAuth client. It then deactivates the user who logged in with a password found to be compromised. You could take other steps here, too. 
+
+You could do more within FusionAuth, by, for example:
+
+* Adding a date of deactivation to the `user.data` field
+* [Actioning the user](/docs/v1/tech/apis/actioning-users) for display in the administrative interface or to be queried via the API
+* Putting the user in a [group](/docs/v1/tech/core-concepts/groups) for future processing
+
+You could also integrate with other systems.
+
+* You could fire off an API call to another service that needs to know about this security violation.
+* The system could add an event to a streaming service, such as Kafka, for future analysis.
+* Your application could send an email to the user and their boss about the situation. Wouldn't be cool, but you could do it.
+
+```php
+//...
+http_response_code(500);
+//...
+```
+
+Finally, the code returns a `500` HTTP status code. This stops the login process. Because you configured this tenant to require all webhooks to succeed before processing the event, if any don't, the event doesn't complete. This means the user is not logged in. 
+
+If this webhook didn't fail, the user would be logged in. The account would be deactivated; they'd be unable to login later. But their current session would be active for the duration of the token. We don't want that to happen, so that's why the code returns a `500`.
+
+## Results
+
+If you install the webhook, follow the instructions in [the repository](https://github.com/FusionAuth/fusionauth-example-php-webhook/blob/master/README.md), and login as a user with a breached password, the user will see this screen on their first failed login: 
```

Comment:
I don't think this message is correct...  assuming this is the error you're seeing when the webhook fails? This may be a bug. I can take a look. 

You don't need to hold up the blog, we can always revise with a new screenshot if needed.

Reply:
Correct, that is the error I'm seeing. Would you like me to file a bug?

Reply:
I just checked , we have a test for this, so not sure how this is happening. You should see something like `One or more webhooks returned an invalid response or were unreachable. Based on your transaction configuration, your action cannot be completed` - and that message is theme-able.